narwhal/primary/src/certificate_waiter.rs

```diff
@@ -3,233 +3,404 @@
+            // Drop sync target when cert store already has an equal or higher round for the origin.
+            // This will apply GC to targets as well.
+            committed_round < target_round
+        });
+        if self.targets.is_empty() {
+            trace!("Certificates have caught up. Skip fetching.");
+            return;
+        }
+
+        let state = self.state.clone();
+        let committee = self.committee.clone();
+
+        debug!(
+            "Starting task to fetch missing certificates: max target {}, committed {:?}",
+            self.targets.values().max().unwrap_or(&0),
+            committed_rounds.values()
+        );
+        self.fetch_certificates_task.push(
+            tokio::task::spawn(async move {
+                state
+                    .metrics
+                    .certificate_waiter_inflight_fetch
+                    .with_label_values(&[&committee.epoch.to_string()])
+                    .inc();
+                state
+                    .metrics
+                    .certificate_waiter_fetch_attempts
+                    .with_label_values(&[&committee.epoch.to_string()])
+                    .inc();
+
+                let now = Instant::now();
+                match run_fetch_task(state.clone(), committee.clone(), committed_rounds).await {
+                    Ok(_) => {
+                        debug!(
+                            "Finished task to fetch certificates successfully, elapsed = {}s",
+                            now.elapsed().as_secs_f64()
+                        );
+                    }
+                    Err(e) => {
+                        debug!("Error from task to fetch certificates: {e}");
+                    }
+                };
+
+                state
+                    .metrics
+                    .certificate_waiter_op_latency
+                    .with_label_values(&[&committee.epoch.to_string()])
+                    .observe(now.elapsed().as_secs_f64());
+                state
+                    .metrics
+                    .certificate_waiter_inflight_fetch
+                    .with_label_values(&[&committee.epoch.to_string()])
+                    .dec();
+            })
+            .boxed(),
+        );
+    }
 
-            self.update_metrics(waiting.len());
+    fn gc_round(&self) -> Round {
+        self.rx_consensus_round_updates
+            .borrow()
+            .to_owned()
+            .saturating_sub(self.gc_depth)
+    }
+
+    fn last_committed_round(&self, authority: &PublicKey) -> DagResult<Round> {
+        match &self.consensus_store {
+            // Last round is 0 (genesis) when authority is not found in store.
+            Some(consensus_store) => Ok(consensus_store
+                .read_last_committed_round(authority)?
+                .unwrap_or(0)),
+            None => Ok(self
+                .certificate_store
+                .last_round_number(authority)?
```

Comment:
For internal consensus if round R is committed for an origin, receiving certificates below R for the same origin becomes no-op because they will not make it into the dag.

For external consensus indeed round R-1 of an origin may be needed later. I will just rely on BlockSynchronizer's causal completion at that point, and will chat with @akichidis to confirm.

Reply:
Added a comment. And in the next PR this logic will unlikely become unnecessary and get removed. The plan is to switch to using a map of written certs above gc_round.

Reply:
> For internal consensus if round R is committed for an origin, receiving certificates below R for the same origin becomes no-op because they will not make it into the dag.

Yes but: a certificate depending on this older certificate might need it in order to qualify for processing, since we still expect all parents to be present to accept it right? @mwtian